crates/vm/src/arch/execution_mode/tracegen/normal.rs

```diff
@@ -0,0 +1,104 @@
+use openvm_instructions::instruction::Instruction;
+use openvm_stark_backend::p3_field::PrimeField32;
+
+use crate::{
+    arch::{
+        execution_control::ExecutionControl, ExecutionError, ExecutionState, InstructionExecutor,
+        VmChipComplex, VmConfig, VmSegmentState,
+    },
+    system::memory::MemoryImage,
+};
+
+pub type TracegenCtx = ();
+
+/// Implementation of the ExecutionControl trait using the old segmentation strategy
+pub struct TracegenExecutionControl {
+    // State
+    pub clk_end: u64,
+    // TODO(ayush): do we need this if only executing one segment?
+    pub final_memory: Option<MemoryImage>,
+}
+
+impl TracegenExecutionControl {
+    pub fn new(clk_end: u64) -> Self {
+        Self {
+            clk_end,
+            final_memory: None,
+        }
+    }
+}
+
+impl<F, VC> ExecutionControl<F, VC> for TracegenExecutionControl
+where
+    F: PrimeField32,
+    VC: VmConfig<F>,
+{
+    type Ctx = TracegenCtx;
+
+    fn should_suspend(
+        &mut self,
+        state: &mut VmSegmentState<Self::Ctx>,
+        _chip_complex: &VmChipComplex<F, VC::Executor, VC::Periphery>,
+    ) -> bool {
+        state.clk >= self.clk_end
+    }
+
+    fn on_start(
+        &mut self,
+        state: &mut VmSegmentState<Self::Ctx>,
+        chip_complex: &mut VmChipComplex<F, VC::Executor, VC::Periphery>,
+    ) {
+        let timestamp = chip_complex.memory_controller().timestamp();
```

Comment:
I think this is always 1, but depends on how we're using memory_controller

Reply:
I'll check